src/typescript-worker/playground-typescript-worker.ts

```diff
@@ -279,30 +290,93 @@ type ResolvedSpecifier = {
 
 /**
  * Resolve an import specifier. Uses HTML-spec semantics for URLs and relative
- * paths, and returns an unpkg.com URL for bare-specifiers.
+ * paths. Uses import map configuration if provided and matching, otherwise
+ * falls back to unpkg.com URL for bare-specifiers.
  */
-const resolveSpecifier = (
-  specifier: string,
-  referrer: string | URL
-): ResolvedSpecifier => {
-  try {
-    return {
-      type: 'url' as const,
-      url: new URL(specifier).href,
-    };
-  } catch (e) {
-    if (isRelativeOrAbsolutePath(specifier)) {
+class ModuleResolver {
+  private importMap: ModuleImportMap;
+
+  constructor(importMap: ModuleImportMap) {
+    this.importMap = importMap;
+  }
+
+  resolve(specifier: string, referrer: string | URL): ResolvedSpecifier {
+    const importMapUrl = this._resolveUsingImportMap(specifier);
+    if (importMapUrl !== null) {
+      return {type: 'bare' as const, url: importMapUrl};
+    }
+    try {
+      return {
+        type: 'url' as const,
+        url: new URL(specifier).href,
+      };
+    } catch (e) {
+      if (isRelativeOrAbsolutePath(specifier)) {
+        return {
+          type: 'relative' as const,
+          url: new URL(specifier, referrer).href,
+        };
+      }
       return {
-        type: 'relative' as const,
-        url: new URL(specifier, referrer).href,
+        type: 'bare' as const,
+        url: `https://unpkg.com/${specifier}`,
       };
     }
-    return {
-      type: 'bare' as const,
-      url: `https://unpkg.com/${specifier}`,
-    };
   }
-};
+
+  private _resolveUsingImportMap(specifier: string): string | null {
+    // For overview, see https://github.com/WICG/import-maps
+    // For algorithm, see https://wicg.github.io/import-maps/#resolving
+    // TODO(aomarks) Add support for `scopes`.
+    for (const [specifierKey, resolutionResult] of Object.entries(
+      this.importMap.imports ?? {}
+    )) {
+      // Note that per spec we shouldn't do a lookup for the exact match case,
+      // because if a trailing-slash mapping also matches and comes first, it
+      // should have precedence.
+      if (specifierKey === specifier) {
+        return resolutionResult;
+      }
+
+      if (specifierKey.endsWith('/') && specifier.startsWith(specifierKey)) {
+        if (!resolutionResult.endsWith('/')) {
+          console.warn(
+            `Could not resolve module specifier "${specifier}"` +
+              ` using import map key "${specifierKey}" because` +
+              ` address "${resolutionResult}" must end in a forward-slash.`
+          );
+          return null;
+        }
+
+        const afterPrefix = specifier.substring(specifierKey.length);
+        let url;
+        try {
+          url = new URL(afterPrefix, resolutionResult);
```

Comment:
will this only happen when resolutionResult isn't a valid base URL? Could it be an early warning too?